File: solutions/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        chars = set()

        left = right = 0
        while right < len(s):
            if not s[right] in chars:
                chars.add(s[right])
                right += 1
                res = max(res, right - left)
            else:
                chars.remove(s[left])
                left += 1

        return res

# @lc code=end

File: solutions/3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int res = 0;
        Set<Character> set = new HashSet<>();

        int left = 0;
        int right = 0;

        while (right < s.length()) {
            if (!set.contains(s.charAt(right))) {
                set.add(s.charAt(right++));
                res = Math.max(res, right - left);
            } else {
                set.remove(s.charAt(left++));
            }
        }

        return res;
    }

    public int lengthOfLongestSubstring_V2(String s) {
        int res = 0;
        Set<Character> set = new HashSet<>();
        int left = 0;
        int right = 0;
        for(char c : s.toCharArray()) {
            right++;
            if (set.contains(c)) {
                while (s.charAt(left) != c)  {
                    set.remove(s.charAt(left++));
                }
                left++;
            } else {
                set.add(c);
            }
            res = Math.max(res, right - left);
        }
        return res;
    }
}
// @lc code=end

